examples/nivc_dynamic_phi.rs

```diff
@@ -0,0 +1,360 @@
+    let preimage = &z[0];
+
+    let preimage_bits = preimage.to_bits_le_strict(cs.namespace(|| "sha_preimage_bits"))?;
+
+    let (preimage_bits, _) = preimage_bits.split_at(8 * NUM_BYTES);
+
+    let preimage_bits = preimage_bits
+      .chunks(8)
+      .flat_map(|chunk| {
+        let mut chunk = chunk.to_vec();
+        chunk.reverse();
+        chunk
+      })
+      .collect::<Vec<_>>();
+
+    let digest_bits = sha256(cs.namespace(|| "sha256"), preimage_bits.as_slice())?;
+
+    let (digest_bits, _) = digest_bits.split_at(8 * NUM_BYTES);
+
+    let next_pc_bit = &digest_bits[7];
+
+    let digest_bits = digest_bits
+      .chunks(8)
+      .flat_map(|chunk| {
+        let mut chunk = chunk.to_vec();
+        chunk.reverse();
+        chunk
+      })
+      .collect::<Vec<Boolean>>();
+
+    let digest = pack_bits(cs.namespace(|| "digest_from_bits"), digest_bits.as_slice())?;
+
+    let new_pc = AllocatedNum::alloc(cs.namespace(|| "next_pc"), || {
+      if self.next_pc {
+        Ok(F::ONE)
+      } else {
+        Ok(F::ZERO)
+      }
+    })?;
+
+    cs.enforce(
+      || "enforce new_pc is first bit",
+      |lc| lc + CS::one(),
+      |lc| lc + new_pc.get_variable(),
+      |_| next_pc_bit.lc(CS::one(), F::ONE),
+    );
+
+    Ok((Some(new_pc), vec![digest]))
+  }
+}
+
+#[derive(Clone, Debug)]
+struct BlakeCircuit<F: PrimeField> {
+  next_pc: bool,
+  _p: PhantomData<F>,
+}
+
+impl<F: PrimeField> BlakeCircuit<F> {
+  fn new(next_pc: bool) -> Self {
+    Self {
+      next_pc,
+      _p: PhantomData,
+    }
+  }
+}
+
+impl<F: PrimeField + PrimeFieldBits> StepCircuit<F> for BlakeCircuit<F> {
+  fn arity(&self) -> usize {
+    1
+  }
+  fn synthesize<CS: ConstraintSystem<F>>(
+    &self,
+    cs: &mut CS,
+    _pc: Option<&AllocatedNum<F>>,
+    z: &[AllocatedNum<F>],
+  ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+    let preimage = &z[0];
+
+    let preimage_bits = preimage.to_bits_le_strict(cs.namespace(|| "blake_preimage_bits"))?;
+
+    let (preimage_bits, _) = preimage_bits.split_at(8 * NUM_BYTES);
+
+    let digest_bits = blake2s(cs.namespace(|| "blake2s"), preimage_bits, b"personal")?;
+
+    let (digest_bits, _) = digest_bits.split_at(8 * NUM_BYTES);
+
+    let next_pc_bit = digest_bits.first();
+
+    let digest = pack_bits(cs.namespace(|| "digest_from_bits"), digest_bits)?;
+
+    let new_pc = AllocatedNum::alloc(cs.namespace(|| "next_pc"), || {
+      if self.next_pc {
+        Ok(F::ONE)
+      } else {
+        Ok(F::ZERO)
+      }
+    })?;
+
+    cs.enforce(
+      || "enforce new_pc is first bit",
+      |lc| lc + CS::one(),
+      |lc| lc + new_pc.get_variable(),
+      |_| next_pc_bit.unwrap().lc(CS::one(), F::ONE),
+    );
+
+    Ok((Some(new_pc), vec![digest]))
+  }
+}
+
+#[derive(Debug)]
+struct ExampleSteps<G1, G2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+{
+  next_pc: bool,
+  _p: PhantomData<(G1, G2)>,
+}
```

Comment:
I don't understand the need for or idea behind the struct. The name implies a collection of steps, yet you're only representing a single bit of information. Moreover, this is 'redundant hint' as described above. And as far as I can tell, this is a second level of indirection that redundant hint — since you're also keeping such hints in the individual hash circuits.